monero/transaction/__init__.py

```diff
@@ -140,27 +140,38 @@ def outputs(self, wallet=None):
         already generated.
         """
 
-        def _scan_pubkeys(svk, psk, stealth_address, amount, encamount):
+        def _scan_pubkeys(svk, psk, stealth_address, amount, encamount, on_chain_vt):
             for keyidx, tx_key in enumerate(self.pubkeys):
                 # precompute
                 svk_2 = ed25519.scalar_add(svk, svk)
                 svk_4 = ed25519.scalar_add(svk_2, svk_2)
                 svk_8 = ed25519.scalar_add(svk_4, svk_4)
                 #
-                hsdata = b"".join(
+                if on_chain_vt:
+                    shared_secret = ed25519.scalarmult(svk_8, tx_key)
+                    vt_hsdata = b"".join([
+                                        b"view_tag",
+                                        shared_secret,
+                                        varint.encode(idx)
+                    ])
+                    vt_full = keccak_256(vt_hsdata).digest()
+                    vt = binascii.hexlify(vt_full)[0:2]
+                    if vt != on_chain_vt:
+                        continue
+                else:
```

Comment:
This shouldn't be an else, the rest of the function should remain unchanged :)

(When scanning, doing the rest of this for every output is slow especially `scalarmult_B`; view tags just short-circuit so it doesn't have to do the rest of this for ~99.6% of outputs, but it still needs to do it to check for ownership when `vt == on_chain_vt`)